paymentapi/src/main/java/fi/hel/verkkokauppa/payment/api/PaymentAdminController.java

```diff
@@ -94,4 +127,39 @@
         }
     }
 
+    @PostMapping(value = "/payment-admin/online/create/card-renewal-payment", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<String> createCardRenewalPayment(@RequestBody GetPaymentRequestDataDto dto) {
+        try {
+
+            Payment payment = service.getPaymentRequestDataForCardRenewal(dto);
+            String vismaPaymentUrl = service.getPaymentUrl(payment);
+            return ResponseEntity.status(HttpStatus.CREATED).body(vismaPaymentUrl);
+        } catch (CommonApiException cae) {
+            throw cae;
+        } catch (Exception e) {
+            log.error("creating payment or card-renewal-payment failed", e);
+            throw new CommonApiException(
+                    HttpStatus.INTERNAL_SERVER_ERROR,
+                    new Error("failed-to-create-card-renewal-payment", "failed to create card renewal payment")
+            );
+        }
+    }
+
+    @GetMapping(value = "/payment-admin/online/cancel", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<VismaPayResponse> cancelPayment(@RequestParam(value = "paymentId") String paymentId) {
+        try {
+            return ResponseEntity.status(HttpStatus.CREATED).body(
+                    cancelPaymentFetcher.cancelPayment(paymentId)
```

Comment:
Just wondering why cancel functionality is not part of the service but separated to own logic/fetcher

Reply:
This is moved to separated Fetcher because this cancels payment on visma.